# Subject: Classical Data Analysis

## Session 2 - Regression with multiple variables

### Exercise 1 -  Linear Regression with multiple variables 

Considering the Stepwise Regression presented in Demo 1 perform the Linear Regression with multiple variable considering the following independent variables:
    
- Situation 1: radio and newspaper;
- Situation 2: TV, radio and newspaper.
    
Interpret and discuss the OLS Regression Results, namely the different R-squared (R2). 

Commit scripts in your GitHub account. You should export your solution code (.ipynb notebook) and push it to your repository “ClassicalDataAnalysis”.

The following are the tasks that should complete and synchronize with your repository “ClassicalDataAnalysis” until November 10. Please notice that none of these tasks is graded, however it’s important that you correctly understand and complete them in order to be sure that you won’t have problems with further assignments.

# 1. Data

Let's take a look again at some data - Advertising - , ask some questions about that data, and then use linear regression (with one and multiple variables) to answer those questions!

In [24]:
# imports
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
%matplotlib inline

In [25]:
# read data into a DataFrame
df = pd.read_csv('http://www-bcf.usc.edu/~gareth/ISL/Advertising.csv', index_col=0)
df.head()

TV  radio  newspaper  sales
1  230.1   37.8       69.2   22.1
2   44.5   39.3       45.1   10.4
3   17.2   45.9       69.3    9.3
4  151.5   41.3       58.5   18.5
5  180.8   10.8       58.4   12.9

- Situation 1: radio and newspaper;
- Situation 2: TV, radio and newspaper.

# 2 - Linear Regression with multiple variables in Statsmodels

## 2.1. Regression model with multiple variables and with a constant for Situation 1

In [12]:
X = df[['radio', 'newspaper']]
y = df['sales']

In [13]:
X= sm.add_constant(X)
model = sm.OLS(y,X).fit()
predictions = model.predict(X)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.333
Model:                            OLS   Adj. R-squared:                  0.326
Method:                 Least Squares   F-statistic:                     49.11
Date:                Fri, 27 Oct 2017   Prob (F-statistic):           4.96e-18
Time:                        12:02:57   Log-Likelihood:                -573.24
No. Observations:                 200   AIC:                             1152.
Df Residuals:                     197   BIC:                             1162.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          9.1889      0.628     14.640      0.000       7.951      10.427
radio          0.1990      0.022      9.101      0.000       0.156       0.242
newspaper      0.0066      0.015      0.446      0.656      -0.023       0.036
==============================================================================
Omnibus:                       19.261   Durbin-Watson:                   1.943
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               21.768
Skew:                          -0.765   Prob(JB):                     1.88e-05
Kurtosis:                       3.519   Cond. No.                         92.0
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## 2.2. Regression model with multiple variables and with a constant for Situation 2

In [14]:
X = df[['TV','radio', 'newspaper']]
y = df['sales']

In [16]:
X = sm.add_constant(X)
model = sm.OLS(y,X).fit()
predictions = model.predict(X)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Fri, 27 Oct 2017   Prob (F-statistic):           1.58e-96
Time:                        12:04:28   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
radio          0.1885      0.009     21.893      0.000       0.172       0.206
newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## 2.3. Regression model with multiple variables and without a constant for Situation 1

In [19]:
X = df[['radio', 'newspaper']]
y = df['sales']

In [20]:
model = sm.OLS(y,X).fit()
predictions = model.predict(X)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.831
Model:                            OLS   Adj. R-squared:                  0.830
Method:                 Least Squares   F-statistic:                     487.9
Date:                Fri, 27 Oct 2017   Prob (F-statistic):           3.02e-77
Time:                        12:07:52   Log-Likelihood:                -646.85
No. Observations:                 200   AIC:                             1298.
Df Residuals:                     198   BIC:                             1304.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
radio          0.3763      0.026     14.334      0.000       0.325       0.428
newspaper      0.1024      0.019      5.303      0.000       0.064       0.140
==============================================================================
Omnibus:                       37.865   Durbin-Watson:                   1.653
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               53.382
Skew:                          -1.137   Prob(JB):                     2.56e-12
Kurtosis:                       4.110   Cond. No.                         3.16
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## 2.4. Regression model with multiple variables and without a constant for Situation 2

In [22]:
X = df[['TV','radio', 'newspaper']]
y = df['sales']

In [23]:
model = sm.OLS(y,X).fit()
predictions = model.predict(X)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.982
Model:                            OLS   Adj. R-squared:                  0.982
Method:                 Least Squares   F-statistic:                     3566.
Date:                Fri, 27 Oct 2017   Prob (F-statistic):          2.43e-171
Time:                        12:08:13   Log-Likelihood:                -423.54
No. Observations:                 200   AIC:                             853.1
Df Residuals:                     197   BIC:                             863.0
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
TV             0.0538      0.001     40.507      0.000       0.051       0.056
radio          0.2222      0.009     23.595      0.000       0.204       0.241
newspaper      0.0168      0.007      2.517      0.013       0.004       0.030
==============================================================================
Omnibus:                        5.982   Durbin-Watson:                   2.038
Prob(Omnibus):                  0.050   Jarque-Bera (JB):                7.039
Skew:                          -0.232   Prob(JB):                       0.0296
Kurtosis:                       3.794   Cond. No.                         12.6
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Interpreting the fitness of the models

In the 1st situation: 

    radio and newspaper 
    
    
    with the constant: Rsqaure 0.333
        p-value newspaper is above 0.05, this variable is not significant
    
    without the constant:  Rsquare 0.831
        p-value


In the 2nd situation:

    TV, radio and newspaper
    
    with the constant: Rsquared 0.897
          the newspaper shows a high p-value,and is not significant
          both radio and TV shows a low p-value and both of these two variable are significant, and thus can be used
          as predictor variables
          
    
    without the constant: Rsquared 0.982
     
     
    The model without the constant is better with a lower rsquared. 